src/nvim/api/ui.c

```diff
@@ -129,15 +129,22 @@ void remote_ui_disconnect(uint64_t channel_id)
 /// Wait until ui has connected on stdio channel.
 void remote_ui_wait_for_attach(void)
 {
-  Channel *channel = find_channel(CHAN_STDIO);
-  if (!channel) {
-    // this function should only be called in --embed mode, stdio channel
-    // can be assumed.
-    abort();
-  }
+  do {
+      uint64_t chan_id = CHAN_LISTEN; 
+      Channel *channel = find_channel(chan_id);
+      if (!channel) {
+          chan_id = CHAN_STDIO;
+          channel = find_channel(chan_id);
+      }
```

Comment:
this looks a bit racy, what if it takes time for the `--listen` connection to be made and it gets stuck waiting on `CHAN_STDIO` ? We should detect the condition of `--embed --listen` from command parsing, and in that case, accept an `nvim_ui_attach` on _any_ channel.

Reply:
What's the cleanest way to pass that info into nvim_ui_attach from your pov? Right now I have it such that nvim_ui_attach always accepts a connection on any channel

Reply:
What I mean is - would it be ok to add an arg to nvim_wait_for_ui_attach or would that be changing the api in an undesirable way. 

Happy to clean up the commit comment and make whatever changes. 

I have a fork of vimr nearly up and running cleanly st. It can talk to a more or less stock nvim with this change.